Renovate the remote local setup

istvanballok opened the pull request:
**How to categorize this PR?**
/area dev-productivity
/kind enhancement

**What this PR does / why we need it**:
This PR contains several enhancements for the "remote local setup" to better support the daily activities of a Gardener developer based on my experience during the past months.

Noteworthy changes:
- Speed up the startup process by using a persistent volume to retain caches and relevant artifacts across restarts, just like running the local setup on your laptop.
- Upgrade the base image and install a few more useful tools.
- Add various minor improvements (bash, tmux configuration and some useful scripts) to make the developer experience more pleasant.
- Support multiple local setup scenarios (gardener, gardener HA, operator, operator complete) in dedicated tmux sessions.

Hopefully with this setup, the entry barrier of e.g. trying out a PR might become even lower than before.

**Which issue(s) this PR fixes**:
N/A

**Special notes for your reviewer**:
/cc @rfranzke, @ScheererJ, @oliver-goetz , @vicwicker, @rickardsjp 
FYI @plkokanov, @seshachalam-yv, @shafeeqes, @acumino, @jguipi, @bd3lage 
FYI @tobschli

To host the remote local setup, I suggest using a Gardener shoot cluster with two worker groups and a daily hibernation schedule.

- `2 CPU, 4 GB,   min: 1, max: 2` - for the system components
- `16 CPU, 32 GB, min: 0, max: n` - for hosting remote local setup pods (1 per node) with compute resources comparable to a modern developer laptop

Multiple developers could share the same shoot cluster and deploy their own remote local setup pods in separate namespaces. Please be invited to use my shoot cluster in the SAP internal dev landscape for a test drive.

To deploy the remote local setup:

```text
$ k create ns <user>
$ k apply -n <user> -f docs/deployment/content/remote-local-setup.yaml
$ k exec -n <user> -it remote-local-setup-0 -- sh
Welcome to the remote local setup pod of Gardener!
...

/ # tmux a
```

**Release note**:

```other developer
The "remote local setup" is enhanced e.g. to support multiple scenarios.
```

istvanballok committed:
Use a persistent volume for the remote local setup

By changing the remote local setup 'deployment' to a 'statefulset' we can use a
persistent volume for the home folder of the remote local setup pod.

The home folder contains valuable artifacts that need not be recreated each
time the remote local setup pod is started. These artifacts include:

- the git clone of the gardener/gardener repository,
- the golang compiler cache,
- the container images that were downloaded by
  - the dind docker daemon (e.g. the KinD image) or
  - containerd inside the KinD cluster or shoot nodes.

    These images are cached by the registry mirror pods in the dev folder of the
    gardener repository.

This allows us to persist this state across restarts and can significantly
reduce the startup time of the local setup scenario in subsequent executions:

- The golang compiler can reuse the compiled artifacts of the packages that did
  not change since the previous build,
- the previously used container images do not need to be downloaded by dockerd
  or containerd from the Internet and
- skaffold can reuse the previously built images and does not need to rebuild
  them.

This makes the remote local setup pod as efficient as a regular local setup
running on a developer's machine, where these artifacts are typically preserved.

To get a fresh state, developers can simply

- scale the remote local setup statefulset to zero replicas,
- delete the persistent volume claim and
- scale the statefulset back to one replica.

Getting a fresh state is typically not necessary as these artifacts are managed
by git/golang/registry and can be safely reused. They are not deleted on the
local machine either.

This change also allows developers to scale the remote local setup statefulset
to zero replicas when they do not need the remote local setup pod, which can
save cloud resources. This allows developers to share a single Gardener shoot
cluster to host multiple remote local setup pods in different namespaces.

- The cluster itself can be hibernated at night, and the individual remote local
  setup statefulsets can be scaled to zero replicas when the specific developer
  does not need the remote local setup pod. This way the cluster autoscaler can
  remove unnecessary nodes and save cloud resources.

- When the developers need the remote local setup pod again, they can scale the
  statefulset back to one replica and continue with their work where they left
  off with all the artifacts and caches still in place.

istvanballok committed:
Update the base image to latest version of Docker Engine: 26.x

https://docs.docker.com/engine/release-notes
https://hub.docker.com/_/docker/tags?page=&page_size=&ordering=&name=26-dind

- docker:24.0.3-dind was based on Alpine v3.18
- docker:26-dind     is  based on Alpine v3.20 (the latest version)

istvanballok committed:
Install (even more) useful tools with apk

+apache2-utils              # htpasswd
 bash
 bash-completion
+bat                        # https://github.com/sharkdp/bat: A cat(1) clone with wings.
+bind-tools                 # dig
+conntrack-tools            # conntrack
+coreutils                  # https://www.gnu.org/software/coreutils/manual/coreutils.html (*)
 curl
+diffutils                  # diff (*)
+dnsmasq                    # dnsmasq, https://en.wikipedia.org/wiki/Dnsmasq
+docker-bash-completion     # bash completion for docker
+findutils                  # find, xargs (*)
 fzf
 g++
+gawk                       # awk (*)
+gcompat                    # GNU C Library compatibility layer for musl
 git
+git-prompt                 # bash prompt integration for Git
 go
+grep                       # grep (*)
+helm                       # helm
+inotify-tools              # inotifywait
 jq
 k9s
 less
 lsof
 make
 mandoc
 mc
+moreutils                  # ts - timestamp input, https://manpages.debian.org/testing/moreutils/ts.1.en.html
 mount
+ncurses                    # clear (*)
+neovim                     # neovim.io
+parallel                   # parallel
 procps
 sed
+strace                     # strace
+tar                        # tar (*)
+tcpdump                    # tcpdump
 tmux
 tmux-doc
+tzdata                     # Timezone data
+util-linux                 # dmesg (*), findmnt, ..., https://pkgs.alpinelinux.org/package/edge/main/x86_64/util-linux
 vim
+wget                       # wget (*)
 yq

*: use the regular version of these programs instead of the
   limited, busybox version that comes with the Alpine image

Note that https://dl-cdn.alpinelinux.org/alpine/v3.20/community is
included in /etc/apk/repositories of the docker:26-dind image,
so some of the packages (like go) are installed from the community
repository.

https://pkgs.alpinelinux.org/packages?name=go&branch=v3.20&repo=community

The mount package in the main repository is fine.

Alpine v3.20/community contains the currently latest version of
k9s: 0.32.4, and a recent version of go: 1.22.3 (latest is 1.22.4, 2024.6.4),
so k9s and go can simply be installed with apk.

istvanballok committed:
Clone the gardener repository to the golang path

When cloning the gardener repository,
use the pre go module style golang path
for compatibility with older code generation tools:

  ~/go/src/github.com/gardener/gardener

istvanballok committed:
Store the dind docker daemon's artifacts on the persistent volume

The KinD image (kindest/node, ~1GB) need not be downloaded
each time from the Internet.

istvanballok committed:
Install (even more) useful tools with wget

Use wget -N (timestamping) instead of curl (-z) to download the files
only if they are newer than the local copy in ~/.cache/wget.

 kind
 kns
+krew              # https://krew.sigs.k8s.io/, the plugin manager for kubectl command-line tool
 ktx
 kube-ps1
 kubectl
 tmux-completion
+yaml2json         # https://github.com/bronze1man/yaml2json, a command line tool convert from yaml to json

istvanballok committed:
Cache the apk artifacts under ~/.cache/apk

If there is no newer version, the same artifacts need not be
downloaded again from the Internet in subsequent executions
because they are cached on the persistent volume.

istvanballok committed:
Install (even more) useful tools with go

istvanballok committed:
Install (even more) useful tools with krew

istvanballok committed:
Let the init script fail if there is a failure in a shell pipe

istvanballok committed:
Show a greeting when exec-ing into the pod with 'k exec -it ... -- sh'

Previously the greeting 'To attach to the gardener tmux session...'
was shown only when bash was executed and it was even
shown inside the tmux session.

Now it is shown when the alpine default shell: ash is executed, so
when we exec into the pod with
  k exec -it remote-local-setup-0 -- sh
and it is not shown when already in a tmux session.

The text of the greeting is a bit more verbose.

istvanballok committed:
Change the PATH

- Go is no longer installed from a manually downloaded binary
  to /usr/local/go/bin, it is installed by apk to /usr/bin/go,
  so a PATH entry is no longer needed for go
- Add GOPATH and the binaries that were installed with 'go install'
  to $GOPATH/bin to the PATH
- The .krew/bin folder needs to be added to the path
  https://krew.sigs.k8s.io/docs/user-guide/setup/install/
- The hack/usage folder is also added to the path for convenience

istvanballok committed:
Show the git prompt and the time in the bash prompt

istvanballok committed:
tmux automatically sets the TERM env variable

to tmux-256color

istvanballok committed:
Retain more entries in the bash history

Furthermore, the commands are appended to the bash history after they are
executed and not only when bash is closed.

Hence all the previously executed commands are visible in the history
of a new bash session.

istvanballok committed:
Use nvim by default, e.g. for k edit

istvanballok committed:
Configure a simple, black and white skin for k9s

In the default MacOS Terminal app, with white background, the default
skin and the alternative skins do not look good / are not really
usable, e.g. the row selection color is too similar to the row text
color, or the prompt suggestions use a barely visible color.

This probably "Keep it simple ..." skin seems to be the best option,
even if not perfect in a white tmux session, e.g. the confirmation
dialog's OK/Cancel buttons are not highlighted when creating a
port-forward.

istvanballok committed:
Disable software flow control

Input XON/XOFF flow control is not needed and this functionality
blocks the keyboard shortcuts: C-S and C-Q.

C-R can be used to search the bash history backward,
and after this commit C-S can search forward.

istvanballok committed:
Preserve the logs of the dind docker daemon

istvanballok committed:
Run dind dockerd and hence the kind cluster with lower priority

This helps to make sure that the interactive processes like tmux, bash
stay responsive even at high load.

istvanballok committed:
Simplify the startup probe definition

With the default periodSeconds being 10s, a failureThreshold of 18
gives 180s or 3 minutes for the startup time.

istvanballok committed:
Increase the resource requests to require a 16 core / 32 GB node

A 16 core and 32 GB k8s node that is expected to run a single instance
of the remote local setup pod has comparable compute resources like a
modern developer laptop.

This capacity is sufficient to run the local setup in a smooth way.

Note that the resource requests are actually smaller than the expected
capacity of the node: 32GB -> 28GB, 16 core -> 14 core.
This reduction is needed to consider the kube reserved capacity: some
CPU and memory is dedicated for the kubelet and other system processes
and those resources are not available for regular pods.

In an example Gardener shoot with default configuration:

k describe node <node>
Capacity:
  cpu:                16
  ephemeral-storage:  64746412Ki
Allocatable:
  cpu:                15920m
  ephemeral-storage:  62985309545

Furthermore, it leaves some place to run daemonsets or other system pods
on the node. These system pods have typically a high priority such that
they would make room for themselves and the remote local setup pod
would be unschedulable if they can't all fit on a node.

The buffer of 2 cores and 4 GB is just an estimate. It is important
that the resource request is more than the size of a next smaller node
(8core/16GB) and that on a single node of the expected size (16core/32GB)
only 1 remote local setup pod would fit.

Note that using even larger nodes where multiple remote local setup pods
could fit is a configuration that I did not test: the recommendation is
to use one node for one remote local setup pod.

istvanballok committed:
Slightly improve the comment on the cgroup bind mount

istvanballok committed:
Use wait to keep the remote-local-setup pod running

Previously reading from stdin was used to block the init script.

The docker daemon is started as a shell background job so it is more
idiomatic to wait for that background job to finish.
The dockerd process is not going to finish so this similarly blocks
the init script indefinitely.

Using this approach, the PID 1 ash shell reaps the children that it
adopts from parents that exited and left some orphan child processes
behind, i.e. this approach prevents the accumulation of zombie processes.

istvanballok committed:
Improve tmux mouse support in selection mode

When in selection mode, use

- the right mouse click to copy the selection to the clipboard and
  leave the selection mode

  the default behaviour was to copy and leave the selection mode after
  the mouse drag finished

  this helps to select text again and again e.g. during a presentation
  without losing the scroll position of the pane

- Command right click is the middle mouse click on Mac OS. It can be
  used to copy the text without leaving selection mode.

istvanballok committed:
Improve tmux behavior when 2 developers collaborate in the same session

When multiple developers attach to the same tmux session, use the smallest
window size.

The default behaviour is latest, such that tmux uses the size of the
client that had the most recent activity. This is convenient for the
active client, but confusing for the passive peer developer as tmux might
show only a fraction of the screen.

For collaboration it is beneficial to align the window size during the
call and using the smallest window size helps to make the window size
difference apparent.

http://man.openbsd.org/OpenBSD-current/man1/tmux.1

istvanballok committed:
Update the /etc/hosts entries

See
- docs/deployment/getting_started_locally.md
- docs/concepts/operator.md

 127.0.0.1 api.e2e-default-wl.local.external.local.gardener.cloud
 127.0.0.1 api.e2e-default-wl.local.internal.local.gardener.cloud
 127.0.0.1 api.e2e-default.local.external.local.gardener.cloud
 127.0.0.1 api.e2e-default.local.internal.local.gardener.cloud
+127.0.0.1 api.e2e-fd-hib.local.external.local.gardener.cloud
+127.0.0.1 api.e2e-fd-hib.local.internal.local.gardener.cloud
+127.0.0.1 api.e2e-force-delete.local.external.local.gardener.cloud
+127.0.0.1 api.e2e-force-delete.local.internal.local.gardener.cloud
 127.0.0.1 api.e2e-hib-wl.local.external.local.gardener.cloud
 127.0.0.1 api.e2e-hib-wl.local.internal.local.gardener.cloud
 127.0.0.1 api.e2e-hib.local.external.local.gardener.cloud
 127.0.0.1 api.e2e-hib.local.internal.local.gardener.cloud
 127.0.0.1 api.e2e-managedseed.garden.external.local.gardener.cloud
 127.0.0.1 api.e2e-managedseed.garden.internal.local.gardener.cloud
-127.0.0.1 api.e2e-managedseed.local.external.local.gardener.cloud
-127.0.0.1 api.e2e-managedseed.local.internal.local.gardener.cloud
 127.0.0.1 api.e2e-mgr-hib.local.external.local.gardener.cloud
 127.0.0.1 api.e2e-mgr-hib.local.internal.local.gardener.cloud
 127.0.0.1 api.e2e-migrate-wl.local.external.local.gardener.cloud
 127.0.0.1 api.e2e-migrate-wl.local.internal.local.gardener.cloud
 127.0.0.1 api.e2e-migrate.local.external.local.gardener.cloud
 127.0.0.1 api.e2e-migrate.local.internal.local.gardener.cloud
 127.0.0.1 api.e2e-rotate-wl.local.external.local.gardener.cloud
 127.0.0.1 api.e2e-rotate-wl.local.internal.local.gardener.cloud
 127.0.0.1 api.e2e-rotate.local.external.local.gardener.cloud
 127.0.0.1 api.e2e-rotate.local.internal.local.gardener.cloud
 127.0.0.1 api.e2e-unpriv.local.external.local.gardener.cloud
 127.0.0.1 api.e2e-unpriv.local.internal.local.gardener.cloud
 127.0.0.1 api.e2e-upd-node-wl.local.external.local.gardener.cloud
 127.0.0.1 api.e2e-upd-node-wl.local.internal.local.gardener.cloud
 127.0.0.1 api.e2e-upd-node.local.external.local.gardener.cloud
 127.0.0.1 api.e2e-upd-node.local.internal.local.gardener.cloud
 127.0.0.1 api.e2e-upg-hib-wl.local.external.local.gardener.cloud
 127.0.0.1 api.e2e-upg-hib-wl.local.internal.local.gardener.cloud
 127.0.0.1 api.e2e-upg-hib.local.external.local.gardener.cloud
 127.0.0.1 api.e2e-upg-hib.local.internal.local.gardener.cloud
 127.0.0.1 api.e2e-upgrade-wl.local.external.local.gardener.cloud
 127.0.0.1 api.e2e-upgrade-wl.local.internal.local.gardener.cloud
 127.0.0.1 api.e2e-upgrade.local.external.local.gardener.cloud
 127.0.0.1 api.e2e-upgrade.local.internal.local.gardener.cloud
 127.0.0.1 api.e2e-wake-up-wl.local.external.local.gardener.cloud
 127.0.0.1 api.e2e-wake-up-wl.local.internal.local.gardener.cloud
 127.0.0.1 api.e2e-wake-up.local.external.local.gardener.cloud
 127.0.0.1 api.e2e-wake-up.local.internal.local.gardener.cloud
 127.0.0.1 api.local.local.external.local.gardener.cloud
 127.0.0.1 api.local.local.internal.local.gardener.cloud
+127.0.0.1 g-seed.ingress.local.seed.local.gardener.cloud
+127.0.0.1 gu-local--local.ingress.local.seed.local.gardener.cloud
+127.0.0.1 p-local--local.ingress.local.seed.local.gardener.cloud
+127.0.0.1 p-seed.ingress.local.seed.local.gardener.cloud
+127.0.0.1 v-local--local.ingress.local.seed.local.gardener.cloud
+127.0.0.3 api.virtual-garden.local.gardener.cloud

istvanballok committed:
Resolve *.seed.local.gardener.cloud to 127.0.0.1 with dnsmasq

in the remote local setup pod.

This is useful when executing the e2e tests locally.
In the CI environment these domains are resolved with additional entries
in /etc/hosts. Dnsmasq allows to map wildcard domains as well,
so this static configuration is more convenient.

See
https://github.com/gardener/gardener/blob/b55d4f13c9a08599d6569b19daf01c49c1dbd436/hack/test-e2e-local.sh#L105-L106

istvanballok committed:
Add bash functions to support a PR test workflow

When testing a PR,
it is helpful to checkout the base commit of the PR first,
start the local setup, inspect its relevant state,
then checkout the head commit of the PR,
upgrade the local setup and
assert the expected change in the relevant state.

The pr and pr_base functions help to checkout the respective git commits.

istvanballok committed:
Use a copy in the default kubeconfig path: ~/.kube/config

Previously, the KUBECONFIG was set to the path of the kubeconfig
that is generated by Gardener's make kind.*-up target.

This leads to issues when using the 'kns' tool: kns edits the kubeconfig
in place and sets the namespace of the current context accordingly.
The skaffold / helm based deployment of gardener fails if the
namespace is set in the kubeconfig that make gardener-up uses.

  make kind-up     # the namespace is not specified in the kubeconfig
  make gardener-up # OK
  kns kube-system  # specify a namespace in the kubeconfig
  make gardener-up # fails with
    waiting for deletion: ...
    You must pass '--namespace=garden' to perform this operation.
    the namespace from the provided object "garden"
    does not match the namespace "kube-system".

It is common to use kubectl alias-es like ks to work in the kube-system
namespace.

To allow to continue using the kns tool, the mitigation is to use
a copy of the Gardener generated kubeconfig in interactive sessions.
This way the copy can be freely edited without impacting the
Gardener deploy process. This is also mentioned in the documentation.

> > It might be helpful to copy this file to `$HOME/.kube/config`,
> > since you will need to target this KinD cluster multiple times.
> docs/deployment/getting_started_locally.md

The sync_kubeconfig function can be executed as a background bash job
in the tmux pane of 'make kind.*-up'. This is more convenient than
copying the regenerated kubeconfig after the kind-up target finished.
sync_kubeconfig uses inotify and copies the kubeconfig after it
changed.

If the sync_kubeconfig runs in the background, after
make kind-down kind-up, the newly generated kubeconfig is
automatically copied to ~/.kube/config

istvanballok committed:
dummy kubeconfig for the initial PS1 prompt

When the remote local setup is started the first time, the kube ps1
prompt is empty in the prepared tmux panes because there is no k8s
cluster yet.

This adds a dummy context 'kind-gardener-local' so that the prompt
looks similar to having a k8s cluster.

istvanballok committed:
Tmux utility functions to facilitate preparing tmux sessions

When sending commands to bash in a tmux pane, we need to wait for bash
to be ready, i.e. to show the prompt.
Afterwards we need to wait for bash to echo back the command.
Otherwise the input can be sent too fast and some of the lines might get
lost (only) visually.

istvanballok committed:
The tmux -u flag is not needed if the language is set to UTF-8

man tmux
-u Write UTF-8 output to the terminal even if the first environment
   variable of LC_ALL, LC_CTYPE, or LANG that is set does not contain
   "UTF-8" or "UTF8".

istvanballok committed:
Tmux session for "make kind-up gardener-up"

istvanballok committed:
Tmux session for "make kind-operator-up operator-up"

istvanballok committed:
Tmux session for "make kind-up gardener-up in High Availability mode"

This tmux session is prepared by default.
Compared to the previous tmux session, it contains several windows
and panes that can help to inspect the state of the local setup.

istvanballok committed:
Use the Berlin time zone as local time

istvanballok committed:
Fix bash completion for kubectl

/etc/bash/bashrc loads /etc/bash/bash_completion.sh which loads
/usr/share/bash-completion/bash_completion

If the file /usr/share/bash-completion/bash_completion is not loaded,
the kubectl completion fails with
  _get_comp_words_by_ref is not defined.

The completion scripts are loaded lazily, so to support completing
the k symlink before completing kubectl, the completion script
needs to be loaded explicitly.

istvanballok committed:
Tmux session for "make kind-operator-up operator-up operator-seed-up"

istvanballok committed:
Use stern to get the logs of multiple pods

github.com/stern/stern

istvanballok committed:
Instead of using a bash alias 'k' for kubectl, use a symlink

This way 'k' can be used in a watch expression, e.g.
  watch k get pods

istvanballok committed:
Specify KUBE_EDITOR for k9s

https://k9scli.io/topics/install

vicwicker commented on docs/deployment/content/remote-local-setup.yaml:
```diff
@@ -1,97 +1,564 @@
+
+          ${y}The purge this environment, delete both the StatefulSet and the pvc:${n}
+            kubectl delete statefulset remote-local-setup
+            kubectl delete pvc    home-remote-local-setup-0
```
Is this tabulation on purpose to align the "remote" words? Otherwise I'd propose to align the arguments:

```
kubectl delete statefulset remote-local-setup
kubectl delete pvc         home-remote-local-setup-0
```

istvanballok replied:
Yes, the purpose was to align on the `remote-local-setup` word and show the `home-` prefix in the blank space that is available after the shorter `pvc` word. I think it looks a bit tidier this way. Could we keep it this way?

vicwicker commented on docs/deployment/content/remote-local-setup.yaml:
```diff
@@ -1,97 +1,564 @@
+          mkdir -p ~/go/src/github.com/gardener
+          cd ~/go/src/github.com/gardener
+          [ -d gardener/.git ] || git clone -q https://github.com/gardener/gardener
+          nice dockerd-entrypoint.sh 2>&1 | tee ~/.dockerd.log &
```
I am not sure if `tee` runs with lower priority in this case (does the pipe come before `nice`?), and I suppose it doesn't make a big difference if it does or not but if the argument to use `nice` is to make sure that the interactive processes like `tmux` or `bash` stay responsive even at high load, then IMHO it would make sense to make sure this argument is also consistent with `tee`

istvanballok replied:
I double checked and it is correct that the nice command only applies to the command immediately following it in a bash pipeline. It does not apply to the entire pipeline. I was actually not aware of this.

```bash
nice sleep 1m | sleep 1m | sleep 1m
```

```text
    PID USER      PR  NI    VIRT    RES    SHR S  %CPU  %MEM     TIME+ COMMAND
  12225 root      20   0    4.7m   4.1m   1.6m S   0.0   0.0   0:00.24      `- -bash                                                                                                                                                                                                         
  53233 root      30  10    6.4m   2.2m   1.8m S   0.0   0.0   0:00.00          `- sleep 1m                                                                                                                                                                                                  
  53234 root      20   0    6.4m   2.2m   1.8m S   0.0   0.0   0:00.00          `- sleep 1m                                                                                                                                                                                                  
  53235 root      20   0    6.4m   2.1m   1.6m S   0.0   0.0   0:00.00          `- sleep 1m
```

However, I'd argue for simplicity that the CPU usage of `ts` which timestamps its stdin or `tee` that copies its stdin to a file is not so significant so I think we could still use the simple pattern that the main process (with CPU load potential) gets a reduced priority and the extra commands that deal with formatting or output redirection run with regular priority.

Background: introducing this pattern to use `nice` for most of the tasks was motivated by the experience that with the local setup running in the background, the golang compiliation during `make gardener-up` or `make generate` have the ability to utilize all the CPU cores of the node, which is good by itself, but it added some extra latency to the tmux experience. Using `nice` especially for these processes helps to keep the tmux session interactive and still allows golang to run the compilation on all the CPU cores in parallel. `ts` and `tee` on the other hand do not really have the potential to generate significant CPU load.

Using nice for `dockerd-entrypoint.sh` is important and meaningful, because all the child processes of the docker daemon inherit the reduced priority, i.e. the KinD cluster containing all the Gardener workload runs with reduced priority. Note that these docker daemon child processes (i.e. the KinD cluster with all the workload) are not restricted by the CPU limits of the pod, due to the `/sys/fs/cgroup` configuration further below. Make generate and other interactive commands are executed in the cgroup of the pod, so they are limited to e.g. 14 cores.

```
[~/go/src/github.com/gardener/gardener (⎈|kind-gardener-local-ha-multi-zone:N/A)] (master) [13:41]# cat /proc/self/cgroup 
0::/kubepods.slice/kubepods-podb18ec09f_0815_4d03_9986_8d518d35e860.slice/cri-containerd-99ae324b932929418bdc9521e962603f67f0f0539bdb266c1d23f33d54e35cc4.scope
[~/go/src/github.com/gardener/gardener (⎈|kind-gardener-local-ha-multi-zone:N/A)] (master) [13:41]# cat /proc/92456/cgroup # a random kubelet process
0::/docker/eb333d5ebe5f50132f299421ed25e639ce68fcc8dfaba3084d62f28c4de8f698/kubelet.slice/kubelet.service
```

vicwicker commented on docs/deployment/content/remote-local-setup.yaml:
```diff
@@ -1,97 +1,564 @@
-          requests: {cpu: 8, memory: 8G}
-          limits:   {cpu: 8, memory: 8G}
+          requests: {cpu: 14, memory: 28G}
+          limits:   {cpu: 14, memory: 28G}
```
Probably I am lacking some pieces here, but commit `b9cba0d` talks about 16 cores and 32 GB nodes. I would appreciate if you could elaborate in such commit's message why this is achieved by using 14 cores and 28 GB.

istvanballok replied:
I amended the commit message. Here is a copy of the text:

> Increase the resource requests to require a 16 core / 32 GB node
> 
> A 16 core and 32 GB k8s node that is expected to run a single instance of the remote local setup pod has comparable compute resources like a modern developer laptop.
> 
> This capacity is sufficient to run the local setup in a smooth way.
> 
> Note that the resource requests are actually smaller than the expected capacity of the node: 32GB -> 28GB, 16 core -> 14 core. This reduction is needed to consider the kube reserved capacity: some CPU and memory is dedicated for the kubelet and other system processes and those resources are not available for regular pods.
> 
> In an example Gardener shoot with default configuration:
> 
> ```bash
> k describe node <node>
> Capacity:
>   cpu:                16
>   ephemeral-storage:  64746412Ki
> Allocatable:
>   cpu:                15920m
>   ephemeral-storage:  62985309545
> ```
> 
> Furthermore, it leaves some place to run daemonsets or other system pods on the node. These system pods have typically a high priority such that they would make room for themselves and the remote local setup pod would be unschedulable if they can't all fit on a node.
> 
> The buffer of 2 cores and 4 GB is just an estimate. It is important that the resource request is more than the size of a next smaller node (8core/16GB) and that on a single node of the expected size (16core/32GB) only 1 remote local setup pod would fit.
> 
> Note that using even larger nodes where multiple remote local setup pods could fit is a configuration that I did not test: the recommendation is to use one node for one remote local setup pod.

vicwicker replied:
Thanks a lot for this detailed explanation!

vicwicker commented on docs/deployment/content/remote-local-setup.yaml:
```diff
@@ -1,97 +1,564 @@
+
+          # Resolve e.g. gu-local--e2e-rotate{-wl}.ingress.$seed_name. ... to 127.0.0.1 for the e2e tests:
+          echo 'address=/.seed.local.gardener.cloud/127.0.0.1' > /etc/dnsmasq.d/gardener.conf
+          sed '1inameserver 127.0.0.1' /etc/resolv.conf > /tmp/resolv.conf && cat /tmp/resolv.conf > /etc/resolv.conf
```
Would it work to use `-i` in `sed` to prepend the line in-place so that we don't need the temporary file?

istvanballok replied:
Unfortunately it does not work - using `sed -i` was also my first attempt.

```bash
sed -i '1ix' /etc/resolv.conf
sed: cannot rename /etc/sedJiEJpg: Resource busy
```

The `/etc/resolv.conf` file can be edited _in place_ and its content can be overwritten using the approach above with `cat ... > /etc/resolv.conf`, but it can not be replaced with a different file: although the `-i` option of `sed` suggests that sed is doing an in place edit:

```text
-i[SUFFIX], --in-place[=SUFFIX]
       edit files in place (makes backup if SUFFIX supplied)
```

actually it creates a temporary file and tries to move it to the original file.

```bash
strace sed -i '1ix' /etc/resolv.conf |& grep rename                                                                                                                                      
rename("/etc/seddHPDNj", "/etc/resolv.conf") = -1 EBUSY (Resource busy)                                                                                                                                                                                                                      
writev(2, [{iov_base="cannot rename /etc/seddHPDNj: Re"..., iov_len=43}, {iov_base=NULL, iov_len=0}], 2cannot rename /etc/seddHPDNj: Resource busy) = 4
```

This is correct to make the edit operation atomic, but it does not work for this specific file because in the kubernetes pod this file path is bind mounted into the container:

```bash
findmnt /etc/resolv.conf
TARGET           SOURCE                                                                                                                                          FSTYPE OPTIONS
/etc/resolv.conf /dev/sda3[/var/lib/containerd/io.containerd.grpc.v1.cri/sandboxes/cd3d3673d3ac4b3951a407f9473997810826e2fe5d9e7fbc19b33b41fe4d9dbc/resolv.conf] ext4   rw,relatime,discard,prjquota,errors=remount-ro,mb_optimize_scan=0
```

vicwicker commented on docs/deployment/content/remote-local-setup.yaml:
```diff
@@ -1,97 +1,564 @@
+          make kind-down kind-ha-multi-zone-down kind-operator-down >/dev/null 2>&1
+          cd - >/dev/null
+          mkdir -p ~/.kube
+          echo '{"current-context": "kind-gardener-local"}' > ~/.kube/config
```
If this is not a valid context, isn't this wording a bit confusing because it's pretty similar to the "real" context `kind-gardener-local-ha-multi-zone`? Maybe `N/A` would achieve the same goal and it makes it clearer that there is no context yet.

istvanballok replied:
Here my intention was to use a fake context that looks like the real context:
the tmux session is prepared in several windows when the KinD cluster is not created yet. The KinD cluster is created when the developer attaches to the tmux session and executes the `make kind-up` command in the first window. It is not possible to "update" the bash prompts in the other windows and showing "N/A" there (which is the default behaviour) is not so nice visually: it suggest that there is an issue with the KinD cluster, although this is just an old prompt, from the time before the KinD cluster was created.

So the hack is to fake having a normal prompt with a dummy context using the same name. Then the other windows look just fine.

The alternative I considered was to execute make kind-up as part of the prepare phase, but it gives more control to the developer to execute these commands manually, so I continued with this approach.

vicwicker commented on docs/deployment/content/remote-local-setup.yaml:
```diff
@@ -1,97 +1,564 @@
+          # Resolve e.g. gu-local--e2e-rotate{-wl}.ingress.$seed_name. ... to 127.0.0.1 for the e2e tests:
+          echo 'address=/.seed.local.gardener.cloud/127.0.0.1' > /etc/dnsmasq.d/gardener.conf
+          sed '1inameserver 127.0.0.1' /etc/resolv.conf > /tmp/resolv.conf && cat /tmp/resolv.conf > /etc/resolv.conf
+          nice dnsmasq --no-daemon --log-debug --log-queries 2>&1 | ts > ~/.dnsmasq.log &
```
Same comment as above regarding the priority of `ts` when using the pipe.

istvanballok replied:
See my response in the thread for the

```
nice dockerd-entrypoint.sh 2>&1 | tee ~/.dockerd.log &
```

line. Although `dnsmasq` is not expected to cause a significant CPU load, I think we could continue to apply this pattern of using nice for the main process and letting the small formatting / redirecting tasks to run with regular priority.

vicwicker reviewed (commented):
Thanks a lot for the all work done here and structuring the change so nicely. I have also been using parts of this PR in the remote local setup in the last months and it has been working very nicely for me.

I would only like to make a couple of suggestions or comments, such as increasing the tmux history to a larger number so that we don't lose the output of executed commands (specially in the `ad-hoc` tmux window):

```
set-option -g history-limit 100000
```

But we can also do this in a different PR.

Other comments you can find them below.

istvanballok committed:
Address PR review: Increase the tmux history limit

to avoid losing the output of the executed commands when scrolling
up in the buffer of a tmux pane.

> Buffers
>
> A configurable history buffer is also maintained for each window.
> By default, up to 2000 lines are kept; this can be altered with the
> history-limit option (see the set-option command above).
>
> history-limit lines
>   Set the maximum number of lines held in window history.
>
> http://man.openbsd.org/OpenBSD-current/man1/tmux.1

So this change increases the relatively small buffer size of 2k lines
to 100k lines. The memory usage of tmux should still be OK, maybe
max ~10MB per tmux pane.

Co-authored-by: Istvan Zoltan Ballok <istvan.zoltan.ballok@sap.com>
Co-authored-by: Victor Herrero Otal <victor.herrero.otal@sap.com>

ScheererJ commented on docs/deployment/content/remote-local-setup.yaml:
```diff
@@ -91,7 +90,16 @@ spec:
       volumes:
       - {name: cgroup,  hostPath: {type: Directory, path: /sys/fs/cgroup}}
       - {name: modules, hostPath: {type: Directory, path: /lib/modules}}
+  volumeClaimTemplates:
+  - apiVersion: v1
+    kind: PersistentVolumeClaim
+    metadata: {name: home}
+    spec:
+      accessModes: [ReadWriteOnce]
+      resources: {requests: {storage: 80Gi}}
+      volumeMode: Filesystem
```
In my experience, something in the docker cache accumulates over time. It got to the point that the pod was no longer starting up as no space was left on the device. For the first few occurrences, I just increased the size of the volume. Last time, I also deleted something. Unfortunately, I do not remember what I deleted, but I guess it was the container images.

Did you see this as well? Is there some mitigation in place now that this no longer happens?

istvanballok replied:
I also observed a full disk quite some time ago, but back then I used a smaller persistent disk (maybe 20GB or 40GB). With the 80GB disk and the current base image version I haven't observed any issues in the last months. Unfortunately I haven't investigated back then what was accumulating assuming it was the same issue that you observed. In my experience the go build cache (`/root/.cache/go-build`), the docker artifacts (`/root/.docker`) and the dev folder in the gardener repository have a significant size but I haven't seen a significant growth in the size of the docker folder.

```
2.7G    /root/.cache/go-build
1.6G    /root/.docker/overlay2
3.9G    /root/.docker/volumes
3.3G    /root/go/src/github.com/gardener/gardener/dev/local-registry
```

So there is no mitigation to prevent the issue that you observed with the accumulation of some docker artifacts. Maybe the newer docker base image version also makes a difference. Unfortunately I recently purged my environments to prepare this PR such that I don't have older environments at the moment that I could investigate.

I'll keep an eye on this aspect.

ScheererJ reviewed (commented):
Awesome very nicely structured improvement. Love the additional tools, especially node-shell, which I was dearly missing compared to a pure local setup.

At times, I used the remote local setup for the garden of an extension setup (see https://github.com/gardener/gardener/blob/master/docs/deployment/getting_started_locally_with_extensions.md). Would you consider adding another session for this as well (similar to the ordinary local, HA and operator setups)? In that scenario, I always faced the issue that the script to get a shoot kubeconfig had the namespace baked in. Would you mind adding another script with an additional parameter or enhance the script to support an optional namespace parameter?

Please note that non of my extension setup related requests are required, but they would be welcome extensions. My concern about the disk filling up over time is more relevant.

Apart from that the pull request already looks very good.

istvanballok reviewed (commented):
Thank you for the review, @vicwicker and @ScheererJ. I tried to respond to the individual threads, added a new commit and amended the commit message of a previous one.

> At times, I used the remote local setup for the garden of an extension setup (see https://github.com/gardener/gardener/blob/master/docs/deployment/getting_started_locally_with_extensions.md). Would you consider adding another session for this as well (similar to the ordinary local, HA and operator setups)? In that scenario, I always faced the issue that the script to get a shoot kubeconfig had the namespace baked in. Would you mind adding another script with an additional parameter or enhance the script to support an optional namespace parameter?

Unfortunately I haven't used the extension setup yet, in fear of leaking cloud resources and for my use case, the various local setups were sufficient. So we could prepare a tmux session in collaboration, based on your best practices, and add it to this PR.

> the script to get a shoot kubeconfig had the namespace baked in

the good news here is that I no longer use a custom script in this version, but the `generate-admin-kubeconf.sh` which is added to the path as well. The https://github.com/gardener/gardener/blob/master/hack/usage/generate-admin-kubeconf.sh script has optional arguments for the namespace and the shoot name.

istvanballok committed:
Add a progress indicator if the remote local setup is being initialised

The reinitialisation of the remote local setup pod takes about 40s.

  $ k delete pod remote-local-setup-0
  pod "remote-local-setup-0" deleted

  $ k get pod -w
  NAME                   READY   STATUS    RESTARTS   AGE
  remote-local-setup-0   0/1     Running   0          1s
  remote-local-setup-0   0/1     Running   0          40s
  remote-local-setup-0   1/1     Running   0          40s

If the developers exec into the pod earlier, and attach to the
tmux session, they might see how the tmux session is being prepared:
windows and panes are opened and the commands are entered, ...

During this time, the tmux session is actually not expected to be used.

This progress indicator now shows to the developers that the
initialisation process is not completed yet.

This improves the usability a bit - otherwise, developers would need to
wait with `k get pods -w` for the pod to become ready to avoid exec-ing
to the pod too early. Now they can exec into to the pod right away.

If something goes wrong, the otherwise blocking progress indicator
can be cancelled with C-c.

istvanballok committed:
Add an example to create a managed seed

This is helpful to check the artifacts that are deployed to a managed
seed.

Note that the KinD cluster is conceptually the merge of the
runtime-garden cluster and a seed cluster, so a managed seed is useful
to check the artifacts of a seed, because it does not contain artifacts
of the runtime-garden cluster.

istvanballok committed:
minor: consistent window ordering in the different tmux sessions

istvanballok committed:
Add gardenctl-v2

gardenctl-v2 is installed and configured to allow to practise using
this tool in the local setup.

Note that gardenctl target seed ... does not work in the gardener HA setup,

  # g target seed local-ha-multi-zone
  Error: failed to get kubeconfig for seed /local-ha-multi-zone:
  failed to get secret garden/local-ha-multi-zone.login:
  secrets "local-ha-multi-zone.login" not found

because indeed this secret is not available in the virtual garden cluster.

g target shoot local works as expected

istvanballok committed:
Fix the color issue of k9s inside tmux

In https://github.com/gardener/gardener/commit/728248f38d409b852aed12653669363e53bedb01
the line
  export TERM=xterm-256color
was added when k9s was installed.

In the commit of this PR: "tmux automatically sets the TERM env variable"
I removed this line.

Although it is true that tmux sets TERM to tmux-256color, but tmux itself
also needs to know that the terminal emulator supports 256 colors, when
it attaches to a session.

Furthermore, although not recommended, the developer could also start
bash outside of tmux: in this case, the TERM will be set to xterm-256color.
If bash is started inside tmux, it will be set by tmux to tmux-256color.

So we need to set TERM in the greeting script that sets the environment
variables for the ash shell.

Not exposing the terminal capabilities (256 colors) to tmux caused
the confusing situation that the colors were mapped to a smaller palette
by tmux and e.g. k9s with its default skin did not really look good.

With this proper color support even inside tmux, we can use the default
skin of k9s and it looks good.

ScheererJ reviewed (approved):
Thanks a lot for the improvements and the detailed explanations.

/lgtm
/approve
